File: oop/03-polymorphism/csharp/CreditAccount.cs
/*
 * Demonstrates: Polymorphism - Override
 * Concept: CreditAccount overrides ProcessWithdrawal to check credit limit
 */

namespace FinanceDomain.Polymorphism;

public class CreditAccount : Account
{
    public decimal CreditLimit { get; }

    public CreditAccount(string accountNumber, string accountHolder, decimal creditLimit)
        : base(accountNumber, accountHolder, 0)
    {
        CreditLimit = creditLimit;
    }

    public override TransactionResult ProcessWithdrawal(decimal amount)
    {
        if (amount <= 0)
            return TransactionResult.Failed("Amount must be greater than zero.");

        decimal currentDebt = Math.Abs(_balance);
        decimal availableCredit = CreditLimit - currentDebt;

        if (amount > availableCredit)
            return TransactionResult.Failed($"Exceeds credit limit. Available credit: R{availableCredit:F2}");

        _balance -= amount;
        return TransactionResult.Succeeded($"Charged R{amount:F2}. Outstanding balance: R{Math.Abs(_balance):F2}");
    }

    public override string GetAccountType() => "Credit Account";

    public decimal AvailableCredit => CreditLimit - Math.Abs(_balance);
}

File: oop/03-polymorphism/typescript/creditAccount.ts
/*
 * Demonstrates: Polymorphism - Override
 * Concept: CreditAccount overrides processWithdrawal to check credit limit
 */

import { Account, TransactionResult } from "./account";

export class CreditAccount extends Account {
    public readonly creditLimit: number;

    constructor(accountNumber: string, accountHolder: string, creditLimit: number) {
        super(accountNumber, accountHolder, 0);
        this.creditLimit = creditLimit;
    }

    override processWithdrawal(amount: number): TransactionResult {
        if (amount <= 0) {
            return TransactionResult.failed("Amount must be greater than zero.");
        }

        const currentDebt = Math.abs(this._balance);
        const availableCredit = this.creditLimit - currentDebt;

        if (amount > availableCredit) {
            return TransactionResult.failed(`Exceeds credit limit. Available credit: R${availableCredit.toFixed(2)}`);
        }

        this._balance -= amount;
        return TransactionResult.succeeded(`Charged R${amount.toFixed(2)}. Outstanding balance: R${Math.abs(this._balance).toFixed(2)}`);
    }

    override getAccountType(): string {
        return "Credit Account";
    }

    get availableCredit(): number {
        return this.creditLimit - Math.abs(this._balance);
    }
}
